source/MissionTimer.cpp

```diff
@@ -0,0 +1,392 @@
+		result.waitTime += Random::Int(randomWaitTime);
+
+	// Validate all the actions attached to the timer, and if they're all valid, instantiate them too.
+	string reason;
+	auto ait = actions.begin();
+	for( ; ait != actions.end(); ++ait)
+	{
+		reason = ait->second.Validate();
+		if(!reason.empty())
+			break;
+	}
+	if(ait != actions.end())
+	{
+		Logger::LogError("Instantiation Error: Timer action \"" + TriggerToText(ait->first)
+			+ "\" uses invalid " + std::move(reason));
+		return result;
+	}
+	for(const auto &[trigger, action] : actions)
+		result.actions[trigger] = action.Instantiate(subs, origin, jumps, payload);
+
+	return result;
+}
+
+
+
+bool MissionTimer::IsOptional() const
+{
+	return optional;
+}
+
+
+
+bool MissionTimer::IsComplete() const
+{
+	return isComplete;
+}
+
+
+
+void MissionTimer::Step(PlayerInfo &player, UI *ui, const Mission &mission)
+{
+	if(isComplete)
+		return;
+
+	const Ship *flagship = player.Flagship();
+	if(!flagship)
+		return;
+	// Don't activate or deactivate the timer if the player is taking
+	// off from a planet or traveling through hyperspace.
+	if(flagship->Zoom() != 1. || flagship->IsHyperspacing())
+		return;
+
+	// Determine if the player meets the activation requirements for this timer.
+	if(!CanActivate(flagship, player))
+	{
+		Deactivate(player, ui, mission);
+		return;
+	}
+
+	// This timer is now active and should advance its counter by 1 tick.
+	// If the full wait time has elapsed, this timer is complete.
+	isActive = true;
+	if(++timeElapsed >= waitTime)
+	{
+		auto it = actions.find(TimerTrigger::TIMEUP);
+		if(it != actions.end())
+			it->second.Do(player, ui, &mission);
+		isComplete = true;
+	}
+}
+
+
+
+bool MissionTimer::CanActivate(const Ship *flagship, const PlayerInfo &player) const
+{
+	// If this timer has no requirements to check, then it should be active.
+	if(!hasRequirements)
+		return true;
+
+	// Does the player's system match the system filter?
+	const System *flagshipSystem = flagship->GetSystem();
+	if((system && flagshipSystem != system) ||
+			(!systems.IsEmpty() && !systems.Matches(flagshipSystem)))
+		return false;
+
+	// Does this timer require that the player is solo (i.e. there are
+	// no escorts in the system with the player)?
+	if(requireSolo)
+	{
+		for(const auto &escort : player.Ships())
+		{
+			// Using GetSystem instead of GetActualSystem so that docked
+			// fighters on the player's flagship don't count against them.
+			if(escort.get() != flagship && !escort->IsParked() && !escort->IsDestroyed()
+					&& escort->GetSystem() == flagshipSystem)
+				return false;
+		}
+	}
+
+	// Does this timer require that the player is idle?
+	if(requireIdle)
+	{
+		// The player can't be sending movement commands.
+		if(!flagship->IsThrusting() && !flagship->IsSteering() && !flagship->IsReversing())
+			return false;
+		// And their ship's velocity must be below the max speed threshold.
+		if(flagship->Velocity().LengthSquared() < idleMaxSpeed)
+			return false;
```

Comment:
Am I crazy or are these the exact opposite of what they should be?

```suggestion
		// The player can't be sending movement commands.
		if(flagship->IsThrusting() || flagship->IsSteering() || flagship->IsReversing())
			return false;
		// And their ship's velocity must be below the max speed threshold.
		if(flagship->Velocity().LengthSquared() > idleMaxSpeed)
			return false;
```